File: sdks/go/providers/smallest_ai/protocol.go
package smallest_ai

import (
	"bytes"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"maps"
	"math"
	"net/http"
	"strings"
	"unicode/utf8"

	out "github.com/speechswitch/client/sdks/go/generated/smallest_ai_output"
	"github.com/speechswitch/client/sdks/go/runtime"
)

type Error struct {
	Message string
	Status  runtime.Optional[int]
	Code    runtime.Optional[string]
}

func (e *Error) Error() string { return e.Message }

type packet struct {
	kind, requestID string
	externalID      runtime.Optional[string]
	audio           []byte
	timestamps      []out.SmallestEnvelopeTimestampsItem
	wordIndex       float64
}

func object(data []byte) (map[string]any, error) {
	if !utf8.Valid(data) || !json.Valid(data) {
		return nil, errors.New("Smallest.ai returned invalid JSON")
	}
	decoder := json.NewDecoder(bytes.NewReader(data))
	decoder.UseNumber()
	var value any
	if err := decoder.Decode(&value); err != nil {
		return nil, errors.New("Smallest.ai returned invalid JSON")
	}
	fields, ok := value.(map[string]any)
	if !ok {
		return nil, errors.New("Invalid Smallest.ai response object")
	}
	return fields, nil
}
func audio(value any) ([]byte, error) {
	encoded, ok := value.(string)
	if !ok {
		return nil, errors.New("Invalid Smallest.ai base64 audio")
	}
	data, err := base64.StdEncoding.Strict().DecodeString(encoded)
	if err != nil || base64.StdEncoding.EncodeToString(data) != encoded {
		return nil, errors.New("Invalid Smallest.ai base64 audio")
	}
	return data, nil
}
func decode(data []byte) (packet, error) {
	var p packet
	fields, err := object(data)
	if err != nil {
		return p, err
	}
	p.kind, _ = fields["status"].(string)
	if p.kind == "error" {
		errorFields := fields
		if nested, present := fields["error"]; present {
			var ok bool
			errorFields, ok = nested.(map[string]any)
			if !ok {
				return p, errors.New("Invalid Smallest.ai error response")
			}
		}
		message, ok := errorFields["message"].(string)
		if !ok {
			return p, errors.New("Invalid Smallest.ai error response")
		}
		failure := &Error{Message: message}
		if raw, present := errorFields["code"]; present {
			code, ok := raw.(string)
			if !ok {
				return p, errors.New("Invalid Smallest.ai error response")
			}
			failure.Code = runtime.Some(code)
		}
		return p, failure
	}
	p.requestID, _ = fields["request_id"].(string)
	if p.requestID == "" {
		return p, errors.New("Invalid Smallest.ai request identity")
	}
	if value, present := fields["external_request_id"]; present {
		external, ok := value.(string)
		if !ok {
			return p, errors.New("Invalid Smallest.ai request identity")
		}
		p.externalID = runtime.Some(external)
	}
	if p.kind == "complete" {
		return p, nil
	}
	payload, ok := fields["data"].(map[string]any)
	if !ok {
		return p, errors.New("Invalid Smallest.ai response object")
	}
	if p.kind == "chunk" {
		p.audio, err = audio(payload["audio"])
		return p, err
	}
	if p.kind != "word_timestamp" {
		return p, errors.New("Unknown Smallest.ai WebSocket status")
	}
	word, wordOK := payload["word"].(string)
	rawIndex, indexOK := payload["id"].(json.Number)
	rawStart, startOK := payload["start"].(json.Number)
	rawEnd, endOK := payload["end"].(json.Number)
	index, indexErr := rawIndex.Float64()
	start, startErr := rawStart.Float64()
	end, endErr := rawEnd.Float64()
	start *= 1000
	end *= 1000
	if !wordOK || !indexOK || !startOK || !endOK || indexErr != nil || startErr != nil || endErr != nil || index < 0 || index > 9007199254740991 || index != math.Trunc(index) || math.IsInf(start, 0) || math.IsInf(end, 0) || start < 0 || end < start {
		return p, errors.New("Invalid Smallest.ai word timestamp")
	}
	p.wordIndex = index
	p.timestamps = []out.SmallestEnvelopeTimestampsItem{{Value: word, StartTimeMs: start, EndTimeMs: runtime.Some(end)}}
	return p, nil
}

func (s *stream) runHTTP() error {
	wire := maps.Clone(s.config.wire)
	wire["text"] = s.config.text
	encoded, err := json.Marshal(wire)
	if err != nil {
		return err
	}
	request, err := http.NewRequestWithContext(s.ctx, "POST", s.config.httpURL, bytes.NewReader(encoded))
	if err != nil {
		return err
	}
	request.Header = s.config.headers.Clone()
	request.Header.Set("Content-Type", "application/json")
	accept := "text/event-stream"
	if s.config.protocol == "http" {
		accept = "audio/wav"
	}
	request.Header.Set("Accept", accept)
	response, err := runtime.OpenResponse(request, s.config.transport)
	if err != nil {
		return err
	}
	s.resourceMutex.Lock()
	closed := s.closed
	if !closed {
		s.body = response.Body
	}
	s.resourceMutex.Unlock()
	if closed {
		response.Body.Close()
		return s.ctx.Err()
	}
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		return &Error{Status: runtime.Some(response.StatusCode), Message: fmt.Sprintf("Smallest.ai returned HTTP %d", response.StatusCode)}
	}
	media := strings.ToLower(strings.TrimSpace(strings.SplitN(response.Header.Get("Content-Type"), ";", 2)[0]))
	if media != "" && ((s.config.protocol == "http" && !strings.HasPrefix(media, "audio/") && media != "application/octet-stream") || (s.config.protocol == "sse" && media != "text/event-stream")) {
		return errors.New("Smallest.ai returned an unexpected content type")
	}
	received := false
	if s.config.protocol == "sse" {
		received, err = s.readSSE(response.Body)
		if err != nil {
			return err
		}
	} else {
		for {
			chunk, err := response.Body.Next(s.ctx)
			if err == io.EOF {
				break
			}
			if err != nil {
				return err
			}
			if len(chunk) > 0 {
				received = true
				if err = s.emit(out.SynthesisItemAsBytes{Value: chunk}); err != nil {
					return err
				}
			}
		}
	}
	if !received {
		return errors.New("Smallest.ai returned no audio")
	}
	return nil
}
func (s *stream) readSSE(body runtime.Input[[]byte]) (bool, error) {
	decoder, err := runtime.NewSSEDecoder(s.config.limit)
	if err != nil {
		return false, err
	}
	defer decoder.Finish()
	received := false
	for {
		chunk, err := body.Next(s.ctx)
		if err == io.EOF {
			return received, errors.New("Smallest.ai SSE ended before completion")
		}
		if err != nil {
			return received, err
		}
		for i, b := range chunk {
			if i%8192 == 0 {
				if err := s.ctx.Err(); err != nil {
					return received, err
				}
			}
			event, err := decoder.Push(b)
			if err != nil {
				return received, err
			}
			if event == nil {
				continue
			}
			fields, err := object([]byte(event.Data))
			if err != nil {
				return received, err
			}
			_, hasError := fields["error"]
			if event.Event == "error" || fields["status"] == "error" || hasError {
				return received, &Error{Message: "Smallest.ai SSE returned an error"}
			}
			done, ok := fields["done"].(bool)
			if !ok || (fields["status"] != "206" && fields["status"] != "200") || done != (fields["status"] == "200") {
				return received, errors.New("Invalid Smallest.ai SSE status")
			}
			if raw, present := fields["audio"]; present {
				data, err := audio(raw)
				if err != nil {
					return received, err
				}
				if len(data) > 0 {
					received = true
					if err = s.emit(out.SynthesisItemAsBytes{Value: data}); err != nil {
						return received, err
					}
				}
			} else if !done {
				return received, errors.New("Smallest.ai SSE chunk omitted audio")
			}
			if done {
				return received, nil
			}
		}
	}
}

File: sdks/rust/src/providers/smallest_ai/protocol.rs
use super::{failure, Error};
use crate::{
    base64,
    generated::{smallest_ai_output::SmallestEnvelopeTimestampsItem, transport::SseMessage},
    http::TransportError,
    json::Raw,
};

pub(super) enum Payload {
    Chunk(Vec<u8>),
    Word {
        index: f64,
        timestamp: SmallestEnvelopeTimestampsItem,
    },
    Complete,
}
pub(super) struct Packet {
    pub request_id: String,
    pub external_id: Option<String>,
    pub payload: Payload,
}
fn audio(value: Raw<'_>) -> Result<Vec<u8>, TransportError> {
    let encoded = value
        .string()
        .map_err(|_| failure("Invalid Smallest.ai base64 audio"))?;
    base64::decode(&encoded)
        .filter(|v| base64::encode(v) == encoded)
        .ok_or_else(|| failure("Invalid Smallest.ai base64 audio"))
}
pub(super) fn decode(text: &str) -> Result<Option<Packet>, TransportError> {
    let raw = Raw::parse_exact(text).map_err(|_| failure("Smallest.ai returned invalid JSON"))?;
    let fields = raw
        .object()
        .map_err(|_| failure("Invalid Smallest.ai response object"))?;
    if fields.len() == 1
        && fields.get("type").and_then(|v| v.string().ok()).as_deref() == Some("pong")
    {
        return Ok(None);
    }
    let status = fields.get("status").and_then(|v| v.string().ok());
    if status.as_deref() == Some("error") {
        let error = fields
            .get("error")
            .copied()
            .unwrap_or(raw)
            .object()
            .map_err(|_| failure("Invalid Smallest.ai error response"))?;
        let message = error
            .get("message")
            .and_then(|v| v.string().ok())
            .ok_or_else(|| failure("Invalid Smallest.ai error response"))?;
        let code = error
            .get("code")
            .map(|v| v.string())
            .transpose()
            .map_err(|_| failure("Invalid Smallest.ai error response"))?;
        return Err(Box::new(Error {
            message,
            code,
            status: None,
        }));
    }
    let request_id = fields
        .get("request_id")
        .and_then(|v| v.string().ok())
        .filter(|v| !v.is_empty())
        .ok_or_else(|| failure("Invalid Smallest.ai request identity"))?;
    let external_id = fields
        .get("external_request_id")
        .map(|v| v.string())
        .transpose()
        .map_err(|_| failure("Invalid Smallest.ai request identity"))?;
    let payload = if status.as_deref() == Some("complete") {
        Payload::Complete
    } else {
        let data = fields
            .get("data")
            .and_then(|v| v.object().ok())
            .ok_or_else(|| failure("Invalid Smallest.ai response object"))?;
        match status.as_deref() {
            Some("chunk") => Payload::Chunk(audio(
                *data
                    .get("audio")
                    .ok_or_else(|| failure("Invalid Smallest.ai base64 audio"))?,
            )?),
            Some("word_timestamp") => {
                let word = data
                    .get("word")
                    .and_then(|v| v.string().ok())
                    .ok_or_else(|| failure("Invalid Smallest.ai word timestamp"))?;
                let index = data
                    .get("id")
                    .and_then(|v| v.number().ok())
                    .ok_or_else(|| failure("Invalid Smallest.ai word timestamp"))?;
                let start = data
                    .get("start")
                    .and_then(|v| v.number().ok())
                    .ok_or_else(|| failure("Invalid Smallest.ai word timestamp"))?
                    * 1000.0;
                let end = data
                    .get("end")
                    .and_then(|v| v.number().ok())
                    .ok_or_else(|| failure("Invalid Smallest.ai word timestamp"))?
                    * 1000.0;
                if !index.is_finite()
                    || !(0.0..=9007199254740991.0).contains(&index)
                    || index.fract() != 0.0
                    || !start.is_finite()
                    || !end.is_finite()
                    || start < 0.0
                    || end < start
                {
                    return Err(failure("Invalid Smallest.ai word timestamp"));
                }
                Payload::Word {
                    index,
                    timestamp: SmallestEnvelopeTimestampsItem {
                        value: word,
                        start_time_ms: start,
                        end_time_ms: Some(end),
                        kind: Default::default(),
                        source: None,
                    },
                }
            }
            _ => return Err(failure("Unknown Smallest.ai WebSocket status")),
        }
    };
    Ok(Some(Packet {
        request_id,
        external_id,
        payload,
    }))
}
pub(super) fn sse(event: SseMessage) -> Result<(bool, Vec<u8>), TransportError> {
    let raw =
        Raw::parse_exact(&event.data).map_err(|_| failure("Smallest.ai returned invalid JSON"))?;
    let fields = raw
        .object()
        .map_err(|_| failure("Invalid Smallest.ai response object"))?;
    let status = fields.get("status").and_then(|v| v.string().ok());
    if event.event == "error" || status.as_deref() == Some("error") || fields.contains_key("error")
    {
        return Err(Box::new(Error {
            message: "Smallest.ai SSE returned an error".into(),
            status: None,
            code: None,
        }));
    }
    let done = fields
        .get("done")
        .and_then(|v| v.boolean().ok())
        .ok_or_else(|| failure("Invalid Smallest.ai SSE status"))?;
    if !matches!(status.as_deref(), Some("200" | "206"))
        || done != (status.as_deref() == Some("200"))
    {
        return Err(failure("Invalid Smallest.ai SSE status"));
    }
    let bytes = match fields.get("audio") {
        Some(value) => audio(*value)?,
        None if done => Vec::new(),
        None => return Err(failure("Smallest.ai SSE chunk omitted audio")),
    };
    Ok((done, bytes))
}
